Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -95,68 +103,170 @@ internal SFSession GetSession(string connStr, SecureString password)
+        internal async Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
         {
             s_logger.Debug("SessionPool::GetSessionAsync");
             if (!_pooling)
-                return NewSessionAsync(connStr, password, cancellationToken);
-            SFSession session = GetIdleSession(connStr);
-            return session != null ? Task.FromResult(session) : NewSessionAsync(connStr, password, cancellationToken);
+                return await NewSessionAsync(connStr, password, _noPoolingSessionCreationTokenCounter.NewToken(), cancellationToken).ConfigureAwait(false);
+            var sessionOrCreateToken = GetIdleSession(connStr);
+            if (sessionOrCreateToken.Session != null)
+            {
+                _sessionPoolEventHandler.OnSessionProvided(this);
+            }
+            return sessionOrCreateToken.Session ?? await NewSessionAsync(connStr, password, sessionOrCreateToken.SessionCreationToken, cancellationToken).ConfigureAwait(false);
         }
 
         internal SFSession GetSession() => GetSession(ConnectionString, Password);
 
         internal Task<SFSession> GetSessionAsync(CancellationToken cancellationToken) =>
             GetSessionAsync(ConnectionString, Password, cancellationToken);
 
-        private SFSession GetIdleSession(string connStr)
+        internal void SetSessionPoolEventHandler(ISessionPoolEventHandler sessionPoolEventHandler)
+        {
+            _sessionPoolEventHandler = sessionPoolEventHandler;
+        }
+        
+        private SessionOrCreationToken GetIdleSession(string connStr)
         {
             s_logger.Debug("SessionPool::GetIdleSession");
             lock (_sessionPoolLock)
             {
-                for (int i = 0; i < _idleSessions.Count; i++)
+                if (_waitingForSessionToReuseQueue.IsAnyoneWaiting())
+                {
+                    s_logger.Debug("SessionPool::GetIdleSession - someone is already waiting for a session, request is going to be queued");
+                }
+                else
                 {
-                    if (_idleSessions[i].ConnectionString.Equals(connStr))
+                    var session = ExtractIdleSession(connStr);
+                    if (session != null)
                     {
-                        SFSession session = _idleSessions[i];
-                        _idleSessions.RemoveAt(i);
-                        long timeNow = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
-                        if (session.IsExpired(_timeout, timeNow))
-                        {
-                            session.close();
-                            i--;
-                        }
-                        else
+                        s_logger.Debug("SessionPool::GetIdleSession - no one was waiting for a session, a session was extracted from idle sessions");
```

Comment:
I would just say it in different words (extracted seems not exact):
"Provided an idle session since new session request queue was empty"

Reply:
done